### DISCUSSION TITLE
Interactive parsing nushell of data structures -- dynamic _data_ browsers

### DISCUSSION THREAD

**Discussion by rudolphpienaar** [nushell/nushell#9325] [Ideas]
# Preamble

Apologies if this has come up... I did look through the issues and discussions (admittedly not a terribly _deep_ dive) and didn't see anything.

# Abstract

The "structured data" model view of nushell offers many interesting ways to think of _data_ in the context of shells and shell scripts. In particular, the tabulated/structured representation of data (esp `json`, `yml`, etc) seems to lend itself naturally to thinking about interactive/dynamic navigation through the data. Imagine being able to "open" a data structure and have the result "block" (i.e. not return to the prompt), but have keyboard controls to "navigate" through the data.

This discussion is merely an invitation/thought experiment about what that might enable in terms of CLI interfaces and what might be needed to implement this.

# Disclaimer

I'm a nu "newbie"  (a "nubie")? I have started using `nushell` more and more, I've written `nushell` scripts to use custom python apps that speak to a REST APIs and find the rendering of `json` responses from these python apps using nu to provide considerable value to users.  

# Introduction

Some REST APIs provide links in the response, in particular `collection+json`  APIs provide responses that explicitly include the space of possible API endpoints available from any particular URL. Returning these responses using `nushell` is particularly attractive to clients from a usability perspective.  The tabulated and organized results provide powerful insights into the data.

While this discussion was/is in the context of such responses, it is not limited only to them. Still, if a url link is contained in a `nushell` cell, being able to "navigate" to this link, and "execute" it, and then have the response drawn in-place, would be a powerful way to navigate various web sites. Similarly, for complex and large `json` structures -- collapsing/expanding cells dynamically would be intuitively useful.

# Methods

It's hard for me with my limited nu experience to think about how to make this possible. While data (responses) can be captured in a variable, how can one "render" that data interactively? I can imagine "printing" the table but then blocking continuously on user input, interpreting user input to show/move a cursor, but don't really know how to start thinking about this in `nushell`.

At the very least, some kind of `ncurses` based movement of the cursor on the screen -- but how would the cursor know/return the underlying `nushell` data element?

# Conclusion / need

If a `nushell` data table (basically the result of any `nushell` operation)  were "interactive", i.e. the prompt does not return, but keyboard controls navigate through the table, with attendant "expand"/"collapse"/"execute" of cells and a "redraw" of the table, then the experience would be a powerful CLI _data_ browser (where _data_ can include of course web requests). This would be a further compelling differentiator of `nushell` and were it possible to do this with minimal work from a developer (i.e. if the shell somehow could provide this) I believe many new compelling use cases of `nushell` would naturally become evident.

**Comment by fdncred** [comment]
We have something kind of like what you're asking with the `explore` command. Try `sys | explore` or `$nu | explore`.
